apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -111,33 +118,44 @@ internal_resolve(Protocol, DNSMessage) ->
     Response = erldns_handler:do_handle(DNSMessage, ?LOCALHOST),
     encode_message(Protocol, Response).
 
--spec(resolve(protocol(), [upstream()], binary(), reply_fun()) -> ok).
-resolve(Protocol, Upstreams, Request, Fun) ->
+-spec(external_resolve(protocol(), [upstream()], binary(), reply_fun(), binary(), zone())-> ok).
+external_resolve(Protocol, Upstreams, Request, Fun, FailMsg, Zone) ->
     Workers =
         lists:map(fun (Upstream) ->
             Pid = start_worker(Protocol, Upstream, Request),
             MonRef = monitor(process, Pid),
             {Upstream, Pid, MonRef}
         end, Upstreams),
-    resolve_loop(Workers, Fun).
+    resolve_loop(Workers, Fun, FailMsg, Zone).
+
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun(), binary(), zone()) -> ok).
+resolve_loop(Workers, Fun, FailMsg, Zone) ->
+    resolve_loop(Workers, Fun, FailMsg, Zone, unresolved).
 
--spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun()) -> ok).
-resolve_loop([], _Fun) ->
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun(), binary(), zone(), forward_status()) -> ok).
+resolve_loop([], _Fun, _Zone, _FailMsg, resolved) ->
     ok;
-resolve_loop(Workers, Fun) ->
+resolve_loop([], Fun, Zone, FailMsg, Status) ->
+    ok = reply(Fun, FailMsg),
```

Comment:
I thought making the resolve_loop responsible to detect the failure on Workers by caring a Status, instead of pattern matching on the function result was easier to understand. Happy to discuss.

Reply:
Let me think more about what else we can do here to improve code readability and to make the logic cleaner.

Reply:
Ok, I've been trying to improve the code and push my work [here](https://github.com/dcos/dcos-net/commit/bddda275968b623ef6ffd22a8933140c14e2de82). However, I didn't test it and I'm sure there is a few mistakes. It addresses some of my comments. Basically, I've added `skip` to `reply_fun/0` type. See code snippets below. Let me know what you think about this approach.

https://github.com/dcos/dcos-net/blob/bddda275968b623ef6ffd22a8933140c14e2de82/apps/dcos_dns/src/dcos_dns_handler.erl#L20-L22

https://github.com/dcos/dcos-net/blob/bddda275968b623ef6ffd22a8933140c14e2de82/apps/dcos_dns/src/dcos_dns_handler.erl#L96-L105

https://github.com/dcos/dcos-net/blob/bddda275968b623ef6ffd22a8933140c14e2de82/apps/dcos_dns/src/dcos_dns_handler.erl#L151-L157